CLI: Add Visual Tests addon install auto-migration when upgrading to 8.0.x

ndelangen opened the pull request:
Closes https://github.com/storybookjs/storybook/issues/26665

## What I did

- I added an auto-migration that adds the `@chromatic-com/storybook` addon for visual regression testing.
- Wrote a test for the `check` method of the above auto-migration.

## Checklist for Contributors

### Testing

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to communicate how to test your changes -->

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

I ran this manually on my machine:

1. I created a sandbox
2. I removed the `'@chromatic-dom/storybook'` entry in `main.ts` under `addons`.
3. I ran `../../code/lib/cli/bin/index.js upgrade`
4. I observed the auto-migration being requested
5. I replied "yes"
6. It ran successfully.

### Documentation

<!-- Please check (put an "x" inside the "[ ]") the applicable items below to indicate which documentation has been updated. -->

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

<!-- CANARY_RELEASE_SECTION -->

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

<!-- CANARY_RELEASE_SECTION -->

ndelangen committed:
add VTA automigration

ndelangen committed:
cleanup

ndelangen committed:
add to index

ndelangen committed:
"It should only run if beforeVersion is < 8.0.x where x is the version we release this in"

ndelangen committed:
Add VTA automigration tests

ndelangen commented on code/lib/cli/src/automigrate/fixes/vta.ts:
```diff
@@ -0,0 +1,60 @@
+
+  prompt() {
+    return dedent`
+      We've detected that you're not yet using the Visual Testing Addon. Would you like to add it?
```
I need help writing a proper sentence here that satisfies the requirement.

ndelangen replied:
@jonniebigodes or @kylegach would either of you be able to assist in writing this text to display to users?

jonniebigodes replied:
We can synthesize this to just explain what it is, we don't need to copy and paste the marketing pitch. More even as the message may get mixed up with all that's going on with other migrations.

joevaugh4n replied:
I think we _do_ need the marketing pitch to a degree because we're literally highlighting the value of the addon! We can definitely condense it though (and the prior feedback was that we shouldn't rely on people understanding what visual testing is, which is a good point). How's this? Much shorter - no mention of Chromatic but I think that's ok:

> **New to Storybook 8**: Storybook's Visual Tests addon helps you catch unintentional changes/bugs in your stories. 
> 
> Install Visual Tests addon in your project? [Y]/n

jonniebigodes replied:
I'm good with that wording @joevaugh4n, and one addition to it, we should provide the link to the documentation wouldn't you agree?

kylegach replied:
I understand the need to be as concise as possible, but this is our `#1` opportunity to help someone make an informed decision here.

I think Joe's edit can work, as long as we add the link back in:

> **New to Storybook 8:** Storybook's Visual Tests addon helps you catch unintentional changes/bugs in your stories.
> 
> Learn more: [https://storybook.js.org/docs/writing-tests/visual-testing](https://storybook.js.org/docs/writing-tests/visual-testing)
> 
> Install Visual Tests addon in your project? [Y]/n

I'd personally feel most comfortable if we're upfront about the Chromatic bit, though.

joevaugh4n replied:
yes i prefer your version @kylegach 

i also am open to adding Chromatic back in. that would look like this: 

> New to Storybook 8: Storybook's Visual Tests addon helps you catch unintentional changes/bugs in your stories. The addon is powered by Chromatic, a cloud-based testing tool developed by Storybook's core team.

> Learn more: https://storybook.js.org/docs/writing-tests/visual-testing

> Install Visual Tests addon in your project? [Y]/n

jonniebigodes replied:
@joevaugh4n That works for me, loved the wording, and we'll need to adjust it to reference the documentation, as Kyle mentioned. I understand your reasoning, @kylegach, about being upfront, but advise against it and leave it as is. If the user wants/needs to learn more, we have the documentation and it's up to him to decide if he wants to use it or not.

valentinpalkovic commented on code/lib/cli/src/automigrate/fixes/vta.ts:
```diff
@@ -0,0 +1,60 @@
+    );
+
+    // @ts-expect-error (user might be upgrading from an older version that still had it)
+    const usesMDX1 = mainConfig?.features?.legacyMdx1 === true || false;
```
I guess this is a left over from some copy pasting?

valentinpalkovic commented on code/lib/cli/src/automigrate/fixes/vta.ts:
```diff
@@ -0,0 +1,60 @@
+  versionRange: ['<8.0.7', '>=8.0.7'],
+
+  async check({ mainConfig }) {
+    const hadAddonInstalled = !!mainConfig?.addons?.find((addon) =>
```
Use the helper function `getAddonNames` from '../helpers/mainConfigFile' to extract the addon names

JReinhold commented on code/lib/cli/src/automigrate/fixes/vta.ts:
```diff
@@ -0,0 +1,60 @@
+  async run({ packageManager, dryRun, mainConfigPath, skipInstall }) {
+    if (!dryRun) {
+      const packageJson = await packageManager.retrievePackageJson();
+      const versionToInstall = getStorybookVersionSpecifier(
+        await packageManager.retrievePackageJson()
+      );
+      await packageManager.addDependencies(
+        { installAsDevDependencies: true, skipInstall, packageJson },
+        [`@storybook/addon-mdx-gfm@${versionToInstall}`]
+      );
```
1. Wrong addon
2. The addon doesn't follow SB's versioning so we can't use that for the version, I guess just install latest?

```suggestion
      await packageManager.addDependencies(
        { installAsDevDependencies: true, skipInstall, packageJson },
        ['@chromatic-com/storybook']
      );
```

ndelangen committed:
resolve review comment

https://github.com/storybookjs/storybook/pull/26766#pullrequestreview-1990988907

ndelangen committed:
resolve review comment

https://github.com/storybookjs/storybook/pull/26766#pullrequestreview-1989730532

ndelangen committed:
resolve review comment

https://github.com/storybookjs/storybook/pull/26766#pullrequestreview-1989726124

ndelangen committed:
Add text provided by @joevaugh4n

ndelangen committed:
Merge branch 'next' into norbert/vta-added-on-upgrade

valentinpalkovic commented on code/lib/cli/src/automigrate/fixes/vta.ts:
```diff
@@ -0,0 +1,56 @@
+    return dedent`
+      New to Storybook 8: Storybook's Visual Tests addon helps you catch unintentional changes/bugs in your stories. The addon is powered by Chromatic, a cloud-based testing tool developed by Storybook's core team.
+
+      Learn more: storybook.js.org/docs/writing-tests/visual-testing
```
```suggestion
      Learn more: ${chalk.yellow('https://storybook.js.org/docs/writing-tests/visual-testing'})
```

JReinhold replied:
something about long URLs getting chopped up and not being copy-pastable when inside `boxen`. So I think we might want to keep it as short as possible.

ghengeveld replied:
FWIW the Chromatic CLI uses `cyan` for links.

ndelangen commented on code/lib/cli/src/automigrate/fixes/vta.ts:
```diff
@@ -0,0 +1,56 @@
+  },
+
+  async run({ packageManager, dryRun, mainConfigPath, skipInstall }) {
+    if (!dryRun) {
```
No, why would it?

ndelangen committed:
Fix review comments

ndelangen committed:
Merge branch 'norbert/vta-added-on-upgrade' of https://github.com/storybookjs/storybook into norbert/vta-added-on-upgrade

ndelangen committed:
cleanup

ndelangen committed:
Merge branch 'next' into norbert/vta-added-on-upgrade
